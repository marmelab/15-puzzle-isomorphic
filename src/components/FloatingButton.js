import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import ClassNames from 'classnames';

import withHover from './withHover';

class FloatingButton extends PureComponent {
    static propTypes = {
        children: PropTypes.arrayOf(PropTypes.element).isRequired,
        color: PropTypes.string,
        hover: PropTypes.bool,
        icon: PropTypes.string,
        pulse: PropTypes.bool,
    };

    static defaultProps = {
        color: 'red',
        icon: 'add',
        pulse: false,
    };

    render() {
        const { children, color, hover, icon, pulse } = this.props;

        const mainBtnClass = ClassNames({
            'btn-floating': true,
            'btn-large': true,
            [color]: true,
            pulse,
        });

        const floatingBtnClass = ClassNames({
            show: hover,
            hide: !hover,
        });

        return (
            <div className="fixed-action-btn">
                <a className={mainBtnClass}>
                    <i className="large material-icons">{icon}</i>
                </a>
                <ul className={floatingBtnClass}>
                    {children.map((child, key) =>
                        React.cloneElement(child, {
                            key,
                            className: `${floatingBtnClass}`,
                        }),
                    )}
                </ul>
            </div>
        );
    }
}

export default withHover(FloatingButton);
