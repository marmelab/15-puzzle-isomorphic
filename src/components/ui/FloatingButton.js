import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import ClassNames from 'classnames';

import withHover from '../withHover';

import Fade from '../transitions/Fade';

class FloatingButton extends PureComponent {
    static propTypes = {
        children: PropTypes.arrayOf(PropTypes.element).isRequired,
        color: PropTypes.string,
        hover: PropTypes.bool,
        icon: PropTypes.string,
        pulse: PropTypes.bool,
    };

    static defaultProps = {
        color: 'red',
        icon: 'add',
        pulse: false,
    };

    render() {
        const { children, color, hover, icon, pulse } = this.props;

        const mainBtnClass = ClassNames(
            'btn-floating',
            'btn-large',
            color,
            pulse,
        );

        const floatingBtnClass = ClassNames({
            showElem: hover,
            hideElem: !hover,
        });

        return (
            <div className="fixed-action-btn">
                <a className={mainBtnClass}>
                    <i className="large material-icons">{icon}</i>
                </a>
                <ul>
                    {children.map((child, key) => (
                        <Fade in={hover} key={key}>
                            {React.cloneElement(child, {
                                className: `${floatingBtnClass}`,
                            })}
                        </Fade>
                    ))}
                </ul>
            </div>
        );
    }
}

export default withHover(FloatingButton);
