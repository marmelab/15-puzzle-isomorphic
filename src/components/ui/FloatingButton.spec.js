import React from 'react';
import renderer from 'react-test-renderer';

import FloatingButton from './FloatingButton';

describe('FloatingButton', () => {
    test('should render FloatingButton without error', () => {
        renderer.create(
            <FloatingButton>
                <p>Child 1</p>
                <p>Child 2</p>
            </FloatingButton>,
        );
    });

    test('should render the FloatingButton correctly', () => {
        const component = renderer.create(
            <FloatingButton>
                <p>Child 1</p>
                <p>Child 2</p>
            </FloatingButton>,
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
