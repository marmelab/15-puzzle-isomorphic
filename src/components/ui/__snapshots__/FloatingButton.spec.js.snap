// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FloatingButton should render the FloatingButton correctly 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onTouchStart={[Function]}
>
  <div
    className="fixed-action-btn"
  >
    <a
      className="btn-floating btn-large red"
    >
      <i
        className="large material-icons"
      >
        add
      </i>
    </a>
    <ul>
      <p
        className="hideElem"
        style={
          Object {
            "opacity": 0,
            "transition": "opacity 375ms ease-in-out",
            "visibility": "hidden",
          }
        }
      >
        Child 1
      </p>
      <p
        className="hideElem"
        style={
          Object {
            "opacity": 0,
            "transition": "opacity 375ms ease-in-out",
            "visibility": "hidden",
          }
        }
      >
        Child 2
      </p>
    </ul>
  </div>
</div>
`;
